File: src/dymaptic.GeoBlazor.Core/Components/MediaInfo.cs
namespace dymaptic.GeoBlazor.Core.Components;

[JsonConverter(typeof(MediaInfoConverter))]
public abstract partial class MediaInfo : MapComponent, IProtobufSerializable<MediaInfoSerializationRecord>
{
    /// <summary>
    ///     Indicates the type of media
    /// </summary>
    public abstract string Type { get; }

    public abstract MediaInfoSerializationRecord ToProtobuf();
}

internal class MediaInfoConverter : JsonConverter<MediaInfo>
{
    public override MediaInfo? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        // check the type property and deserialize to the correct type
        var jsonDoc = JsonDocument.ParseValue(ref reader);
        string? type = jsonDoc.RootElement.GetProperty("type").GetString();

        switch (type)
        {
            case "bar-chart":
                return JsonSerializer.Deserialize<BarChartMediaInfo>(jsonDoc.RootElement.GetRawText(), options);
            case "column-chart":
                return JsonSerializer.Deserialize<ColumnChartMediaInfo>(jsonDoc.RootElement.GetRawText(), options);
            case "pie-chart":
                return JsonSerializer.Deserialize<PieChartMediaInfo>(jsonDoc.RootElement.GetRawText(), options);
            case "line-chart":
                return JsonSerializer.Deserialize<LineChartMediaInfo>(jsonDoc.RootElement.GetRawText(), options);
            case "image-media":
                return JsonSerializer.Deserialize<ImageMediaInfo>(jsonDoc.RootElement.GetRawText(), options);
        }

        return null;
    }

    public override void Write(Utf8JsonWriter writer, MediaInfo value, JsonSerializerOptions options)
    {
        var newOptions = new JsonSerializerOptions(options)
        {
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
        };
        writer.WriteRawValue(JsonSerializer.Serialize(value, typeof(object), newOptions));
    }
}

File: src/dymaptic.GeoBlazor.Core/Scripts/mediaInfo.ts
import {buildDotNetBarChartMediaInfo, buildJsBarChartMediaInfo} from "./barChartMediaInfo";
import {buildDotNetColumnChartMediaInfo, buildJsColumnChartMediaInfo} from "./columnChartMediaInfo";
import {buildDotNetImageMediaInfo, buildJsImageMediaInfo} from "./imageMediaInfo";
import {buildDotNetLineChartMediaInfo, buildJsLineChartMediaInfo} from "./lineChartMediaInfo";
import {buildDotNetPieChartMediaInfo, buildJsPieChartMediaInfo} from "./pieChartMediaInfo";

export function buildDotNetMediaInfo(mediaInfo: any): any {
    switch (mediaInfo?.type) {
        case 'bar-chart':
            return buildDotNetBarChartMediaInfo(mediaInfo);
        case 'column-chart':
            return buildDotNetColumnChartMediaInfo(mediaInfo);
        case 'image-media':
            return buildDotNetImageMediaInfo(mediaInfo);
        case 'line-chart':
            return buildDotNetLineChartMediaInfo(mediaInfo);
        case 'pie-chart':
            return buildDotNetPieChartMediaInfo(mediaInfo);
        default:
            throw new Error("Unknown media info type");
    }
}

export function buildJsMediaInfo(dotNetMediaInfo: any): any {
    switch (dotNetMediaInfo?.type) {
        case "bar-chart":
            return buildJsBarChartMediaInfo(dotNetMediaInfo);
        case "column-chart":
            return buildJsColumnChartMediaInfo(dotNetMediaInfo);
        case "image-media":
            return buildJsImageMediaInfo(dotNetMediaInfo);
        case "line-chart":
            return buildJsLineChartMediaInfo(dotNetMediaInfo);
        case "pie-chart":
            return buildJsPieChartMediaInfo(dotNetMediaInfo);
        default:
            throw new Error("Unknown media info type");
    }
}
